pkg/source/examples/simple_source/impl/simple_source.go

```diff
@@ -0,0 +1,88 @@
+package impl
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	sourcesdk "github.com/numaproj/numaflow-go/pkg/source"
+	"github.com/numaproj/numaflow-go/pkg/source/model"
+)
+
+// SimpleSource is a simple source implementation.
+type SimpleSource struct {
+	readIdx  int64
+	toAckSet map[int64]struct{}
+	lock     *sync.Mutex
+}
+
+func NewSimpleSource() *SimpleSource {
+	return &SimpleSource{
+		readIdx:  0,
+		toAckSet: make(map[int64]struct{}),
+		lock:     new(sync.Mutex),
+	}
+}
+
+// Pending returns the number of pending records.
+func (s *SimpleSource) Pending(_ context.Context) uint64 {
+	// The simple source always returns 0 to indicate no pending records.
+	return 0
+}
+
+// Read reads messages from the source and sends the messages to the message channel.
+// If the read request is timed out, the function returns without reading new data.
+// Right after reading a message, the function marks the offset as to be acked.
+func (s *SimpleSource) Read(_ context.Context, readRequest sourcesdk.ReadRequest, messageCh chan<- model.Message) {
+	// Handle the timeout specification in the read request.
+	ctx, cancel := context.WithTimeout(context.Background(), readRequest.TimeOut())
+	defer cancel()
+
+	// If we have un-acked data, we return without reading any new data.
+	// TODO - we should have a better way to handle this.
+	// In real world, there can be the case when the numa container restarted before acknowledging a batch,
+	// leaving the toAckSet not empty on the UDSource container side.
+	// In this case, for the next batch read, we should read the data from the last acked offset instead of returning.
+	// Our built-in Kafka source follows this logic.
```

Comment:
@vigith Thanks for reviewing the PR. I have addressed all of your comments. One more change is that I added this TODO here to tackle the numa container restart case. I will address it in a separate PR.